core/src/main/java/org/apache/iceberg/data/GenericRecord.java

```diff
@@ -28,16 +28,16 @@
 import org.apache.iceberg.StructLike;
 import org.apache.iceberg.relocated.com.google.common.base.Objects;
 import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
-import org.apache.iceberg.relocated.com.google.common.collect.Maps;
 import org.apache.iceberg.types.Types;
 import org.apache.iceberg.types.Types.StructType;
+import org.apache.iceberg.util.CaseInsensitiveMap;
 
 public class GenericRecord implements Record, StructLike {
   private static final LoadingCache<StructType, Map<String, Integer>> NAME_MAP_CACHE =
       Caffeine.newBuilder()
       .weakKeys()
       .build(struct -> {
-        Map<String, Integer> idToPos = Maps.newHashMap();
+        Map<String, Integer> idToPos = new CaseInsensitiveMap<>();
```

Comment:
Could we just use case insensitive strings as keys instead of creating a new map implementation?
e.g. simply using a treemap with a comparator `Map<String, Integer> idToPos = new TreeMap<>(String.CASE_INSENSITIVE_ORDER);` could work I think

Reply:
Make sense, I will consider using your suggestion after this commit can be accept.

Reply:
There is precedence elsewhere for using the treemap comparator as mentioned by @marton-bod.

The times that we already have `CaseInsensitiveMap` are usually using the `CaseInsensitiveMap` from spark (where it wouldn't make sense to use TreeMap and a Comparator in such a way in Scala).

If we do add a `CaseInsensitiveMap`, could it potentially be made a little simpler, maybe even just a utility that returns the above (though I agree, the above inlined is the canonical java answer, such as when I google it etc... there are some scattered implementations but not in any libraries that we already shade such as guava).

Reply:
got it